src/WeightVectors.jl

```diff
@@ -212,15 +215,65 @@ Base.iszero(w::AbstractWeightVector) = w.m[2] == 5
     sample_within_level(rng, m, pos, len)
 end
 
-@inline function sample_within_level(rng, m, pos, len)
+@inline function sample_within_level(rng, m::Memory{UInt64}, pos::UInt64, len::UInt64)
+    shift = leading_zeros(len-1)
+
+    # These should never happen but is required to make `@inbounds` safe:
+    1 <= pos::UInt64 <= UInt64(2)^63 || @fail 0
+    2 <= shift::Int || @fail 0
+    2*(typemax(UInt)>>shift)+pos+1 <= lastindex(m) || @fail 0
+
     while true
-        r = rand(rng, UInt64)
-        k1 = (r>>leading_zeros(len-1))
-        k2 = _convert(Int, k1<<1+pos)
+        r = rand(rng, UInt64)::UInt64
+        k1 = r>>shift
+        k2 = 2k1+pos
```

Comment:
```suggestion
        k2 = _convert(Int, 2k1+pos)
```

Seems there is maybe a slight regression from #163 in the bulk sampling, maybe it is that the conversion actually helps performance? Although the conversion should be accounted for in the proof, nothing changes on the checks you need as far as I understand

Reply:
I was hoping to remove it without performance impact to simplify the proof. But if there's a performance impact of using `Int(x)` instead of `x%Int`, then I'll keep it and add a few clauses to the proof.

Reply:
Seems that it actually improved performance by looking at the new results. Moreover, I checked `@code_native` of `sample_within_level` with `convert` and without and the code is very different and longer when there is no `convert`. So this helps.

Reply:
In particular there is this in the no convert case:

```julia
; │ @ /home/bob/.julia/dev/WeightVectors/src/WeightVectors.jl:229 within `sample_within_level`
; │┌ @ boot.jl:892 within `Int64`
; ││┌ @ boot.jl:817 within `toInt64`
; │││┌ @ boot.jl:756 within `check_sign_bit`
	movabs	rcx, offset jl_small_typeof
	mov	rdi, rax
	mov	rbx, qword ptr [rcx + 256]
	movabs	rcx, offset ijl_box_uint64
	mov	qword ptr [rbp - 56], rbx
	vzeroupper
	call	rcx
	movabs	rcx, offset ".Ljl_sym#convert#4910.jit"
	mov	qword ptr [rbp - 64], rax
	movabs	rdi, offset ".Ljl_global#4909.jit"
	lea	rsi, [rbp - 104]
	mov	edx, 3
	mov	qword ptr [rbp - 104], rcx
	mov	qword ptr [rbp - 96], rbx
	mov	qword ptr [rbp - 88], rax
	movabs	rax, offset j1_throw_inexacterror_4908
	call	rax
	ud2
```

Reply:
Ah yeah sorry, yes, I mean `_convert` and `DEBUG=false`